core/src/main/scala/org/http4s/StructuredField.scala

```diff
@@ -0,0 +1,278 @@
+  */
+sealed trait StructuredField extends Renderable
+
+object StructuredField {
+
+  // 3 top-level types: SfItem, SfList, SfDictionary
+
+  // bare items
+
+  sealed trait BareItem extends Renderable
+
+  object BareItem {
+    val parser: Parser[BareItem] =
+      Rfc8941.bareItem(
+        SfInteger.parser,
+        SfDecimal.parser,
+        SfString.parser,
+        SfToken.parser,
+        SfBinary.parser,
+        SfBoolean.parser,
+      )
+  }
+
+  sealed abstract case class SfInteger(value: Long) extends BareItem {
+    override def render(writer: Writer): writer.type =
+      writer << value
+  }
+
+  object SfInteger {
+    private val MaxValue: Long = 999999999999999L
+
+    private val MinValue: Long = -MaxValue
+
+    private[http4s] def unsafeFromLong(l: Long): SfInteger =
+      new SfInteger(l) {}
+
+    def fromLong(l: Long): Option[SfInteger] =
+      if (l < MinValue || l > MaxValue) None else Some(unsafeFromLong(l))
+
+    def fromInt(i: Int): SfInteger =
+      unsafeFromLong(i.toLong)
+
+    val parser: Parser[SfInteger] =
+      Rfc8941.sfInteger.map(unsafeFromLong)
+  }
+
+  sealed abstract case class SfDecimal(value: BigDecimal) extends BareItem {
+    override def render(writer: Writer): writer.type =
+      if (value.scale > 3)
+        writer << value.setScale(3, BigDecimal.RoundingMode.HALF_UP).toString
+      else if (value.scale < 1)
+        writer << value.setScale(1).toString
+      else
+        writer << value.toString
+  }
+
+  object SfDecimal {
+    private val MaxValue: BigDecimal = BigDecimal("999999999999.999")
+
+    private val MinValue: BigDecimal = -MaxValue
+
+    private[http4s] def unsafeFromBigDecimal(d: BigDecimal): SfDecimal =
+      new SfDecimal(d) {}
+
+    def fromBigDecimal(d: BigDecimal): Option[SfDecimal] =
+      if (d < MinValue || d > MaxValue) None else Some(unsafeFromBigDecimal(d))
+
+    val parser: Parser[SfDecimal] =
+      Rfc8941.sfDecimal.map(unsafeFromBigDecimal)
+  }
+
+  sealed abstract case class SfString(value: String) extends BareItem {
+    override def render(writer: Writer): writer.type =
+      writer << value
+  }
+
+  object SfString {
+    private[http4s] def unsafeFromString(s: String): SfString =
+      new SfString(s) {}
+
+    def fromString(s: String): Option[SfString] =
+      parser.parseAll(s).toOption
+
+    val parser: Parser[SfString] =
+      Rfc8941.sfString.map(unsafeFromString)
+  }
+
+  sealed abstract case class SfToken(value: String) extends BareItem {
+    override def render(writer: Writer): writer.type =
+      writer << value
+  }
+
+  object SfToken {
+    private[http4s] def unsafeFromString(s: String): SfToken =
+      new SfToken(s) {}
+
+    def fromString(s: String): Option[SfToken] =
+      parser.parseAll(s).toOption
+
+    val parser: Parser[SfToken] =
+      Rfc8941.sfToken.map(unsafeFromString)
+  }
+
+  sealed abstract case class SfBinary(value: String) extends BareItem {
```

Comment:
Shouldn't we wrap a byte array here (or something equivalent like `scodec.ByteVector`) rather than the base64 encoded value?

If we *do* store base64, I think we should definitely validate it fully when instantiating/parsing.